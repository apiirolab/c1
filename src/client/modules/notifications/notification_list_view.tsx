import * as React from 'react';

import { inject, observer } from 'mobx-react';
import { CSSTransitionGroup } from 'react-transition-group';
import { Button, Header, List, Message } from 'semantic-ui-react';

import { QueryTypes } from 'data/client';
import { NotificationView } from './notification_item_view';
import { RemoveNotificationsMutation } from './notifications_queries';

type Props = {
  notifications: QueryTypes.NotificationsQuery_Notifications[];
  context?: App.Context;
};

let notification: QueryTypes.NotificationsQuery_Notifications;

export const NotificationsListView: React.SFC<Props> = ({ notifications, context }) => (
  <>
    <Choose>
      <When condition={notifications && notifications.length > 0}>
        <>
          <Header content="Notifications" icon="bell" as="h5" dividing />
          <List relaxed divided>
            <CSSTransitionGroup
              component={React.Fragment}
              transitionName="example"
              transitionAppear={true}
              transitionAppearTimeout={500}
              transitionEnter={false}
              transitionEnterTimeout={500}
              transitionLeave={true}
              transitionLeaveTimeout={500}
            >
              <For each="notification" of={notifications}>
                <NotificationView key={notification.id} notification={notification} />
              </For>
            </CSSTransitionGroup>
          </List>
          <RemoveNotificationsMutation>
            {clearNotifications => (
              <Button
                onClick={() => clearNotifications()}
                color="red"
                icon="trash"
                labelPosition="left"
                content={context.i18n`Clear All`}
              />
            )}
          </RemoveNotificationsMutation>
        </>
      </When>
      <Otherwise>
        <Message>You have no new notifications.</Message>
      </Otherwise>
    </Choose>
  </>
);

NotificationsListView.displayName = 'NotificationsListView';

export const NotificationsListContainer = inject('context')(observer(NotificationsListView));
