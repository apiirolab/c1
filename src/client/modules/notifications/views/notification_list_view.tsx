import * as React from 'react';
import { Header, List, ListProps, Message, Segment } from 'semantic-ui-react';
import styled, { StyledComponentClass } from 'styled-components';
import { NotificationView } from './notification_item_view';


const NotificationList: StyledComponentClass<ListProps, {}> = styled(List)`
  &&&&& .item {
    line-height: inherit;
  }
`;

type Props = {
  notifications: Corpix.Entities.Notification[]
}


let notification: Corpix.Entities.Notification;

export const NotificationListView: React.SFC<Props> = ({ notifications }) => (
  <>
    <Choose>
      <When condition={notifications.length > 0}> 
        <Segment>
          <Header content="Notifications" icon="bell" as="h5" />
           <NotificationList divided>
            <For each="notification" of={notifications}>
              <NotificationView notification={notification} />
            </For>
          </NotificationList> 
        </Segment>
       </When>
      <Otherwise>
        <Message>You have no new notifications.</Message>
      </Otherwise>
    </Choose> 
  </>
);
