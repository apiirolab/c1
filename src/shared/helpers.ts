interface Indexable<T> {
  [index: string]: T;
}

interface Group<T> {
  key: string;
  values: T[];
}

export function groupBy<T>(xs: T[], key: string): Indexable<T[]> {
  return xs.reduce(function(previous: any, current: Indexable<any>) {
    // Indexable<Array<Indexable<T>>>
    (previous[current[key]] = previous[current[key]] || []).push(current);
    return previous;
  }, {});
}

export function groupByArray<T>(xs: T[], key: string | Function): Array<Group<T>> {
  return xs.reduce(function(previous, current: Indexable<any>) {
    let v = key instanceof Function ? key(current) : current[key];
    let el = previous.find(r => r && r.key === v);
    if (el) {
      el.values.push(current);
    } else {
      previous.push({
        key: v,
        values: [current]
      });
    }
    return previous;
  }, []);
}
