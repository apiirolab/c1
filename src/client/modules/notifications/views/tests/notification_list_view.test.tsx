import * as React from 'react';

import { create as render } from 'react-test-renderer';

import { create } from 'shared/test_data';
import { NotificationListView } from '../notification_list_view';


describe('Notifications', () => {
  storyOf(
    'Item View',
    {
      notifications: [
        create.notification({ comment: 'Notification 1', date: new Date(2010, 1, 2, 12, 50)} ),
        create.notification({ comment: 'Notification 2', date: new Date(2010, 1, 2, 3, 50)} ),
        create.notification({ comment: 'Notification 3', date: new Date(2007, 1, 2, 10, 50)} ),
        create.notification({ comment: 'Notification 4', date: new Date(2010, 1, 2, 11, 50)} ),
        create.notification({ comment: 'Notification 5', date: new Date(2011, 1, 2, 12, 50)} )
      ],
      componentWithData(notifications: Corpix.Entities.Notification[]) {
        return <NotificationListView notifications={notifications} />;
      }
    },
    data => {
      it('renders no notifications', () => {
        const root = render(data.componentWithData([]));
        expect(root).toMatchSnapshot();
      });

      it('renders some notifications', () => {
        const root = render(data.componentWithData(data.notifications));
        expect(root).toMatchSnapshot();
      });
    }
  );
})


