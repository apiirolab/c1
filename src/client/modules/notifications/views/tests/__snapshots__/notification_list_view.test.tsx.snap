// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications Item View renders no notifications 1`] = `
<div
  className="ui message"
>
  You have no new notifications.
</div>
`;

exports[`Notifications Item View renders some notifications 1`] = `
<div
  className="ui segment"
>
  <h5
    className="ui header"
  >
    <i
      aria-hidden="true"
      className="bell icon"
    />
    <div
      className="content"
    >
      Notifications
    </div>
  </h5>
  <div
    className="ui divided list sc-bwzfXH flOgBY"
    role="list"
  >
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
    >
      <span
        className="sc-bdVaJa cWCCxy"
      >
        Tue, 02 Feb 2010 01:50:00 GMT
      </span>
       · 
      Notification 1
    </div>
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
    >
      <span
        className="sc-bdVaJa cWCCxy"
      >
        Mon, 01 Feb 2010 16:50:00 GMT
      </span>
       · 
      Notification 2
    </div>
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
    >
      <span
        className="sc-bdVaJa cWCCxy"
      >
        Thu, 01 Feb 2007 23:50:00 GMT
      </span>
       · 
      Notification 3
    </div>
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
    >
      <span
        className="sc-bdVaJa cWCCxy"
      >
        Tue, 02 Feb 2010 00:50:00 GMT
      </span>
       · 
      Notification 4
    </div>
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
    >
      <span
        className="sc-bdVaJa cWCCxy"
      >
        Wed, 02 Feb 2011 01:50:00 GMT
      </span>
       · 
      Notification 5
    </div>
  </div>
</div>
`;
